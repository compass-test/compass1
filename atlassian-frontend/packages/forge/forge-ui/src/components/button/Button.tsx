import React, { useState, useContext } from 'react';
import { ForgeDoc } from '@atlassian/forge-ui-types';
import { RenderFn } from '../../renderer/RendererNext';
import { buttonShouldFitContainer } from './breakpoints';
import { WidthContext } from '../../context';
import Icon from '../icon';

export const AkButton = React.lazy(
  () => import('@atlaskit/button/custom-theme-button'),
);

type PropsOf<T extends (props: any) => React.ReactNode> = Parameters<T>[0];

export function useButton({
  forgeDoc,
  dispatch,
}: Parameters<RenderFn>[0]): {
  akButtonProps: PropsOf<typeof AkButton>;
} {
  const {
    props: {
      appearance = 'default',
      text,
      disabled,
      onClick,
      icon,
      iconPosition = 'before',
    } = {},
  } = forgeDoc;
  const [loading, setLoading] = useState(false);

  const handleClick = async () => {
    setLoading(true);
    try {
      await dispatch({
        type: 'event',
        handler: onClick,
        args: [],
        extensionData: {},
      });
    } finally {
      setLoading(false);
    }
  };

  let iconDom: React.ReactChild | undefined;
  if (icon) {
    iconDom = <Icon size="medium" glyph={icon} />;
  }

  const iconBefore = iconPosition === 'before' ? iconDom : undefined;
  const iconAfter = iconPosition === 'after' ? iconDom : undefined;
  const width = useContext(WidthContext);
  let shouldFitContainer = buttonShouldFitContainer(width);

  return {
    akButtonProps: {
      shouldFitContainer,
      iconBefore,
      iconAfter,
      isLoading: loading,
      isDisabled: disabled,
      appearance,
      onClick: handleClick,
      children: text,
    },
  };
}

export function Button(props: Parameters<RenderFn>[0]) {
  const { akButtonProps } = useButton(props);
  return <AkButton {...akButtonProps} />;
}

export function useActionButtons({
  forgeDoc,
  dispatch,
}: Omit<Parameters<RenderFn>[0], 'forgeDoc'> & { forgeDoc: ForgeDoc[] }): {
  akActionButtonProps: Array<PropsOf<typeof AkButton>>;
} {
  const [loading, setLoading] = useState(forgeDoc.map(() => false));
  const width = useContext(WidthContext);
  const shouldFitContainer = buttonShouldFitContainer(width);

  const actionButtons = forgeDoc.map((actionButtonForgeDoc, index) => {
    const {
      props: {
        appearance = 'subtle',
        text,
        disabled,
        onClick,
        icon,
        iconPosition = 'before',
      } = {},
    } = actionButtonForgeDoc;

    const handleClick = async () => {
      setLoading(
        loading.map((l, i) => {
          if (i === index) {
            return true;
          }
          return l;
        }),
      );
      try {
        await dispatch({
          type: 'event',
          handler: onClick,
          args: [],
          extensionData: {},
        });
      } finally {
        setLoading(
          loading.map((l, i) => {
            if (i === index) {
              return false;
            }
            return l;
          }),
        );
      }
    };

    let iconDom: React.ReactChild | undefined;
    if (icon) {
      iconDom = <Icon size="medium" glyph={icon} />;
    }

    const iconBefore = iconPosition === 'before' ? iconDom : undefined;
    const iconAfter = iconPosition === 'after' ? iconDom : undefined;

    return {
      shouldFitContainer,
      iconBefore,
      iconAfter,
      isLoading: loading[index],
      isDisabled: disabled,
      appearance,
      onClick: handleClick,
      children: text,
    };
  });

  return {
    akActionButtonProps: actionButtons,
  };
}
