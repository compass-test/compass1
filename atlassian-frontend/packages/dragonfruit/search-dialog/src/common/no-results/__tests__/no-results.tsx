import React, { ReactElement } from 'react';
import { NoResults, NoPreQueryResults } from '../';
import Button from '@atlaskit/button';
import {
  shallowWithIntl,
  mountWithIntl,
} from '../../../__tests__/__fixtures__/intl-test-helpers';

const shallow = <P extends {}>(
  node: ReactElement<P>,
  additionalOptions: {} = {},
) => {
  return shallowWithIntl(node, additionalOptions).dive();
};

describe('<NoResults />', () => {
  const advancedSearchSelected = jest.fn();
  const clearFilters = jest.fn();

  const commonProps = {
    advancedSearchUrl: 'advanced/search/url',
    linkComponent: () => <div />,
    advancedSearchSelected,
    clearFilters,
    hasFilters: false,
  };

  it('render should match snapshot', () => {
    const wrapper = shallow(<NoResults {...commonProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('with filters', () => {
    it('render should match snapshot', () => {
      const wrapper = shallow(<NoResults {...commonProps} hasFilters />);
      expect(wrapper).toMatchSnapshot();
    });

    it('on clear filter button press should execute the clearFilters callback', () => {
      const wrapper = mountWithIntl(<NoResults {...commonProps} hasFilters />);
      wrapper.find('EmptyState').find(Button).prop('onClick')!(
        {} as any,
        {} as any,
      );
      expect(clearFilters).toBeCalledTimes(1);
    });
  });
});

describe('<NoPreQueryResults />', () => {
  it('render should match snapshot', () => {
    const wrapper = shallow(<NoPreQueryResults />);
    expect(wrapper).toMatchSnapshot();
  });
});
