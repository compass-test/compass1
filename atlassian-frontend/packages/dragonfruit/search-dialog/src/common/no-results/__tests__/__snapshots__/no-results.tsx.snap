// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoPreQueryResults /> render should match snapshot 1`] = `
<Styled(div)>
  <AsyncComponent />
  <Styled(div)>
    Search for what you need
  </Styled(div)>
  <Styled(div) />
</Styled(div)>
`;

exports[`<NoResults /> render should match snapshot 1`] = `
.c0 {
  padding: 100px;
}

<div
  className="c0"
>
  <EmptyState
    Image={[Function]}
    content={
      <span
        onClick={[Function]}
        role="none"
      >
        <I18nTemplatedLink
          href="advanced/search/url"
          i18nTemplateString="Try again with a different term."
          linkComponent={[Function]}
        />
      </span>
    }
    title="We couldn't find anything matching your search."
  />
</div>
`;

exports[`<NoResults /> with filters render should match snapshot 1`] = `
.c0 {
  padding: 100px;
}

<div
  className="c0"
>
  <EmptyState
    Image={[Function]}
    content={
      <React.Fragment>
        <FormattedMessage
          defaultMessage="Looks like you have some filters applied"
          description="Message to show when there are no search results due to filters being applied"
          id="product_search_dialog.confluence.filters_applied"
          tagName="div"
          values={Object {}}
        />
        <styled.div>
          <Memo(Button)
            appearance="primary"
            onClick={[MockFunction]}
          >
            <FormattedMessage
              defaultMessage="Clear filters"
              description="Message to be displayed on button to clear the filters applied"
              id="product_search_dialog.confluence.clear_filters"
              values={Object {}}
            />
          </Memo(Button)>
        </styled.div>
      </React.Fragment>
    }
    title="We couldn't find anything matching your search."
  />
</div>
`;
